pageserver/src/tenant/timeline.rs

```diff
@@ -4176,6 +4177,398 @@ impl Timeline {
             _ = self.cancel.cancelled() => {}
         )
     }
+
+    /// Detach this timeline from its ancestor by copying all of ancestors layers as this
+    /// Timelines layers up to the ancestor_lsn.
+    ///
+    /// Requires a timeline that:
+    /// - has an ancestor to detach from
+    /// - the ancestor does not have an ancestor -- follows from the original RFC limitations, not
+    /// a technical requirement
+    /// - has prev_lsn in remote storage (temporary restriction)
+    ///
+    /// After the operation has been started, it cannot be canceled. Upon restart it needs to be
+    /// polled again until completion.
+    ///
+    /// During the operation all timelines sharing the data with this timeline will be reparented
+    /// from our ancestor to be branches of this timeline.
+    pub(crate) async fn prepare_to_detach_from_ancestor(
+        self: &Arc<Timeline>,
+        tenant: &crate::tenant::Tenant,
+        options: AncestorDetachOptions,
+        ctx: &RequestContext,
+    ) -> Result<PreparedTimelineDetach, DetachFromAncestorError> {
+        use DetachFromAncestorError::*;
+
+        if self.remote_client.as_ref().is_none() {
+            unimplemented!("no new code for running without remote storage");
+        }
+
+        let Some((ancestor, ancestor_lsn)) = self
+            .ancestor_timeline
+            .as_ref()
+            .map(|tl| (tl.clone(), self.ancestor_lsn))
+        else {
+            return Err(NoAncestor);
+        };
+
+        if !ancestor_lsn.is_valid() {
+            return Err(NoAncestor);
+        }
+
+        if ancestor.ancestor_timeline.is_some() {
+            // non-technical requirement; we could flatten N ancestors just as easily but we chose
+            // not to
+            return Err(TooManyAncestors);
+        }
+
+        if self.get_prev_record_lsn() == Lsn::INVALID
+            || self.disk_consistent_lsn.load() == ancestor_lsn
+        {
+            // this is to avoid a problem that after detaching we would be unable to start up the
+            // compute because of "PREV_LSN: invalid".
+            return Err(DetachedTimelineNeedsWrites);
+        }
```

Comment:
I was thinking of the following:
- store the "wal disaster recovery path" (or "lineage") on `index_part.json`
    - so far we've discussed last 100 `(TimelineId, Lsn)` pairs
    - if you go over 100 timeline detach ancestor, we lose wal DR, no big deal
- use existence of that value and matching lsn on check here: https://github.com/neondatabase/neon/blob/ab95942fc25fa1c6bfd6f3041f16a868e8d86dcf/pageserver/src/basebackup.rs#L604

I didn't want to do index_part.json changes here, because those need a week.